### PR TITLE
[feature-wip](multi-catalog) iceberg v2

### PR DESCRIPTION
**Pull request by wsjz**
# Proposed changes

Support iceberg v2 feature:
When read iceberg mor table, doris will filter rows in delele file of iceberg v2，just read the rows that have not been deleted.

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [x] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [x] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.05 seconds
 load time: 571 seconds
 storage size: 17154644860 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221102124710_clickbench_pr_38569.html

File: be/src/vec/exec/scan/vfile_scanner.cpp [hunk 1]
```diff
@@ -468,18 +471,29 @@ Status VFileScanner::_get_next_reader() {
...
-            init_status = ((ParquetReader*)(_cur_reader.get()))
-                                  ->init_reader(_colname_to_value_range, _push_down_expr);
+            init_status = parquet_reader->init_reader(_file_col_names, _colname_to_value_range,
+                                                      _push_down_expr);
+            if (_params.__isset.table_format_params &&
+                _params.table_format_params.table_format_type == "iceberg") {
+                IcebergTableReader* iceberg_reader = new IcebergTableReader(
+                        (GenericReader*)parquet_reader, _profile, _state, _params);
+                iceberg_reader->init_row_filters();
+                iceberg_reader->filter_rows();
```
Comment (morningman) [inline be/src/vec/exec/scan/vfile_scanner.cpp:492]: `filter_rows()` method should move into `init_row_filters`

File: be/src/vec/exec/format/table/iceberg_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        }
+        Status st = _cur_delete_file_reader->get_next_block(&block, &read_rows, &eof);
+        if (!st.ok() || eof) {
+            if (!_delete_file_readers.empty()) {
+                _cur_delete_file_reader = _delete_file_readers.front();
+                _delete_file_readers.pop();
+            }
+        }
+        if (read_rows != 0) {
+            auto& pos_type_column = block.get_by_name("pos");
```
Comment (morningman) [inline be/src/vec/exec/format/table/iceberg_reader.cpp:80]: "pos" should be define somewhere.

File: be/src/vec/exec/format/parquet/vparquet_reader.cpp [hunk 3]
```diff
@@ -299,6 +312,72 @@ Status ParquetReader::get_columns(std::unordered_map<std::string, TypeDescriptor
     return Status::OK();
 }
 
+void ParquetReader::merge_delete_row_ranges(const std::vector<RowRange>& delete_row_ranges) {
+    //    std::vector<RowRange> group_delete_row_ranges;
```
Comment (morningman) [inline be/src/vec/exec/format/parquet/vparquet_reader.cpp:316]: Remove unused code

File: be/src/vec/exec/format/parquet/vparquet_column_reader.h [hunk 4]
```diff
@@ -106,12 +106,12 @@ class ParquetColumnReader {
...
                                     ColumnSelectVector& select_vector, size_t batch_size,
                                     size_t* read_rows, bool* eof) = 0;
-    static Status create(FileReader* file, FieldSchema* field, const ParquetReadColumn& column,
-                         const tparquet::RowGroup& row_group, std::vector<RowRange>& row_ranges,
+    static Status create(FileReader* file, FieldSchema* field, const tparquet::RowGroup& row_group,
                          cctz::time_zone* ctz, std::unique_ptr<ParquetColumnReader>& reader,
                          size_t max_buf_size);
     void init_column_metadata(const tparquet::ColumnChunk& chunk);
     void add_offset_index(tparquet::OffsetIndex* offset_index) { _offset_index = offset_index; }
+    void add_row_ranges(const std::vector<RowRange>* row_ranges) { _row_ranges = row_ranges; };
```
Comment (morningman) [inline be/src/vec/exec/format/parquet/vparquet_column_reader.h:114]:
```suggestion
    void set_row_ranges(const std::vector<RowRange>* row_ranges) { _row_ranges = row_ranges; };
```

File: be/src/vec/exec/format/parquet/vparquet_reader.cpp [hunk 5]
```diff
@@ -355,25 +438,34 @@ Status ParquetReader::_init_row_group_readers() {
...
 }
 
-Status ParquetReader::_filter_row_groups() {
-    if (_total_groups == 0 || _t_metadata->num_rows == 0 || _range_size < 0) {
+Status ParquetReader::_filter_row_groups(const bool& enabled,
+                                         std::vector<RowGroupIndex>& group_indexes) {
+    if (enabled && (_total_groups == 0 || _t_metadata->num_rows == 0 || _range_size < 0)) {
         return Status::EndOfFile("No row group need read");
     }
+    int32_t start_row_id = 0;
```
Comment (morningman) [inline be/src/vec/exec/format/parquet/vparquet_reader.cpp:446]: if `enabled` is false, you can return immediately.

File: be/src/vec/exec/format/parquet/vparquet_reader.h [hunk 6]
```diff
@@ -53,24 +53,33 @@ class ParquetReader : public GenericReader {
...
+                       const bool& filter_groups = true) {
+        // without predicate
+        return init_reader(column_names, nullptr, nullptr, filter_groups);
+    }
+
     Status init_reader(
+            const std::vector<std::string>& column_names,
             std::unordered_map<std::string, ColumnValueRangeType>* colname_to_value_range,
-            VExprContext* vconjunct_ctx);
+            VExprContext* vconjunct_ctx, const bool& filter_groups = true);
```
Comment (morningman) [inline be/src/vec/exec/format/parquet/vparquet_reader.h:73]:
```suggestion
            VExprContext* vconjunct_ctx, bool filter_groups = true);
```

File: be/src/vec/exec/format/parquet/vparquet_reader.cpp (hunk 5 above)
Comment (wsjz) [reply be/src/vec/exec/format/parquet/vparquet_reader.cpp:446]: whether it is enabled or not，we will generate group index for read in parallel or other optimizations

**Review by morningman — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

File: be/src/vec/exec/format/parquet/vparquet_reader.cpp [hunk 7]
```diff
@@ -372,20 +447,24 @@ bool ParquetReader::_next_row_group_reader() {
...
+        group_index++;
+        RETURN_IF_ERROR(_process_page_index(row_group));
+        if (_row_ranges.empty()) {
+            _row_ranges.emplace_back(0, row_group.num_rows);
             _statistics.read_rows += row_group.num_rows;
         }
-        RETURN_IF_ERROR(row_group_reader->init(_file_metadata->schema(), candidate_row_ranges,
-                                               _col_offsets));
+        RETURN_IF_ERROR(row_group_reader->init(_file_metadata->schema(), _col_offsets));
+        row_group_reader->set_row_ranges(_row_ranges);
```
Comment (AshinGau) [inline be/src/vec/exec/format/parquet/vparquet_reader.cpp:467]: 是不是有正确性问题呢，所有的RowGroupReader都是相同的 _row_ranges

Comment (AshinGau) [reply be/src/vec/exec/format/parquet/vparquet_reader.cpp:467]: 底层的ColumnReader需要的row range是针对自己的行号，每个ColumnReader的行号都是从0开始的。

**Comment by mongo360** [comment]
这个功能什么时候支持 EQUALITY_DELETES 有计划么？

**Pull request merged**